ouroboros-network/src/Ouroboros/Network/Diffusion.hs

```diff
@@ -144,32 +168,97 @@ runDataDiffusion tracers
-
-                  -- If any other threads throws 'cleanNetowrkStateThread' and
-                  -- 'cleanLocalNetworkStateThread' threads will will finish.
-                  Async.waitEither_ cleanNetworkStateThread cleanLocalNetworkStateThread
+    -- Define how to fulfill node-to-client and node-to-node connection
+    -- requests. The GADTs `LocalRequest` and `Request` describe all of the
+    -- possibilities. In particular, the node-to-client protocol is restricted
+    -- to responder only, i.e. this node will never try to initiate a
+    -- node-to-client protocol.
+    let localConnectionRequest
+          :: LocalRequest provenance
+          -> SomeVersionedApplication NodeToClientProtocols NodeToClientVersion DictVersion provenance
+        localConnectionRequest ClientConnection = SomeVersionedResponderApp
+          (NetworkServerTracers
+            dtMuxLocalTracer
+            dtHandshakeLocalTracer
+            dtErrorPolicyTracer)
+          (daLocalResponderApplication applications)
+
+        -- Note-to-node connections request: for a PeerConnection we do the
+        -- responder app `daResponderApplication`, and for the others we choose
+        -- the `daInitiatorApplication`. The types leave little room for error.
+        connectionRequest
+          :: Request provenance
+          -> SomeVersionedApplication NodeToNodeProtocols NodeToNodeVersion DictVersion provenance
+        connectionRequest PeerConnection = SomeVersionedResponderApp
+          (NetworkServerTracers
+            dtMuxTracer
+            dtHandshakeTracer
+            dtErrorPolicyTracer)
+          (daResponderApplication applications)
+        -- IP or DNS subscription requests are locally-initiated (requests
+        -- from subscribers to _us_ are `PeerConnection` above.
+        connectionRequest IpSubscriptionConnection  = SomeVersionedInitiatorApp
+          (NetworkConnectTracers dtMuxTracer dtHandshakeTracer)
+          (daInitiatorApplication applications)
+        connectionRequest DnsSubscriptionConnection = SomeVersionedInitiatorApp
+          (NetworkConnectTracers dtMuxTracer dtHandshakeTracer)
+          (daInitiatorApplication applications)
+
+        -- How to make requests for the accept loops for the local and remote
+        -- server.
+        localClassifyRequest :: SockAddr sockType -> LocalRequest Remote
+        localClassifyRequest _ = ClientConnection
+
+        classifyRequest :: SockAddr sockType -> Request Remote
+        classifyRequest _ = PeerConnection
+
+        acceptException :: Socket.SockAddr -> IOException -> IO ()
+        acceptException a e = do
+          traceWith (WithAddr a `contramap` dtErrorPolicyTracer) $ ErrorPolicyAcceptException e
+
+        -- How to run a local server: take the `daLocalAddress` and run an
+        -- accept loop on it against the node-to-client connetions (yet to
+        -- be constructed).
+        runLocalServer n2cConnections = withSockType addr $ \addr' ->
```

Comment:
Alex, could you annotate this with its type. I think it is something like `Connections ... -> IO ...`